File: ruby/ch1/ex110.rb
require 'test/unit'

module Ex110
  def a(x, y)
    case y
    when 0
      0
    when 1
      2
    else
      x == 0 ? 2 * y : a(x - 1, a(x, y - 1))
    end
  end

  def f(n)
    a(0, n)
  end

  def g(n)
    a(1, n)
  end

  def h(n)
    a(2, n)
  end

  def k(n)
    5 * n * n
  end
end


class MyTest < Test::Unit::TestCase
  include Ex110

  def test_1
    assert_equal(1024, a(1, 10))
    assert_equal(65536, a(2, 4))
    assert_equal(65536, a(3, 3))
    assert_equal(8, f(4))
    assert_equal(156, f(78))
    assert_equal(16, g(4))
    assert_equal(65536, h(4))
  end
end

File: java/src/ch1/ex110.java
package ch1;

public class ex110 {
    public static void main(String[] args) {
        System.out.println("f(4) = " + f(4));
        System.out.println("g(4) = " + g(4));
        System.out.println("h(4) = " + h(4));
    }

    static int A(int x, int y) {
        if (y == 0) return 0;
        else if (x == 0) return 2 * y;
        else if (y == 1) return 2;
        else return A(x - 1, A(x, y - 1));
    }

    static int f(int n) {
        return A(0, n);
    }


    static int g(int n) {
        return A(1, n);
    }


    static int h(int n) {
        return A(2, n);
    }


    static int k(int n) {
        return 5 * n * n;
    }

}
